File: python/debug/decode_custom_error.py
from pathlib import Path
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from sdk import EZManagerSDK


def main() -> int:
    if len(sys.argv) < 2:
        print('Usage: python debug/decode_custom_error.py <revertDataHex>')
        return 2

    data = sys.argv[1].strip()

    # Decoder logic uses local ABI set; no RPC/private key required.
    sdk = EZManagerSDK.__new__(EZManagerSDK)
    from utils import load_abi_map
    sdk.abi = load_abi_map()
    sdk.error_selector_map = EZManagerSDK._build_error_selector_map(sdk)

    decoded = EZManagerSDK.decode_custom_error(sdk, data)
    print(decoded)
    return 0


if __name__ == '__main__':
    raise SystemExit(main())

File: js/debug/decodeCustomError.js
#!/usr/bin/env node
import { ethers } from 'ethers';
import { loadAbiMap } from '../utils.js';

const input = process.argv[2];
if (!input) {
  console.error('Usage: node debug/decodeCustomError.js <revertDataHex>');
  process.exit(2);
}

const data = String(input).trim();
const abi = loadAbiMap();
const interfaces = [
  new ethers.Interface(abi.CL_MANAGER),
  new ethers.Interface(abi.CL_CORE),
  new ethers.Interface(abi.DEX_ADAPTER),
  new ethers.Interface(abi.REBALANCE_PLANNER),
  new ethers.Interface(abi.VALUATION),
  new ethers.Interface(abi.ERC20),
  new ethers.Interface(abi.UNI_FACTORY),
  new ethers.Interface(abi.SLIP_FACTORY),
  new ethers.Interface(abi.UNI_POOL),
  new ethers.Interface(abi.SLIP_POOL)
];

let found = null;
for (const iface of interfaces) {
  try {
    const decoded = iface.parseError(data);
    if (!decoded) continue;
    const signature = decoded.fragment?.format('sighash') || `${decoded.name}()`;
    const names = (decoded.fragment?.inputs || []).map((i) => i.name).filter(Boolean);
    const args = {};
    for (let i = 0; i < names.length; i++) args[names[i]] = decoded.args?.[i];
    found = { name: decoded.name, signature, args };
    break;
  } catch (_) {}
}

if (!found) {
  const selector = data.startsWith('0x') ? data.slice(0, 10) : `0x${data.slice(0, 8)}`;
  console.log(JSON.stringify({ selector, signature: null, args: null }, null, 2));
  process.exit(0);
}

console.log(JSON.stringify(found, null, 2));
